common/djangoapps/entitlements/api/v1/views.py

```diff
@@ -57,3 +64,167 @@ def perform_destroy(self, instance):
             )
         if save_model:
             instance.save()
+
+
+class EntitlementEnrollmentViewSet(viewsets.GenericViewSet):
+    """
+    Endpoint in the Entitlement API to handle the Enrollment of a User's Entitlement.
+    This API will handle
+        - Enroll
+        - Unenroll
+        - Switch Enrollment
+    """
+    authentication_classes = (JwtAuthentication, SessionAuthentication,)
+    permission_classes = (permissions.IsAuthenticated,)
+    queryset = CourseEntitlement.objects.all()
+    serializer_class = CourseEntitlementSerializer
+
+    def _verify_course_run_for_entitlement(self, entitlement, course_run_id):
+        """
+        Verifies that a Course run is a child of the Course assigned to the entitlement.
+        """
+        course_run_valid = False
+        course_runs = get_course_runs_for_course(entitlement.course_uuid)
+        for run in course_runs:
+            if course_run_id == run.get('key', ''):
+                course_run_valid = True
+                break
+        return course_run_valid
+
+    def _enroll_entitlement(self, entitlement, course_run_key, user):
+        """
+        Internal method to handle the details of enrolling a User in a Course Run.
+        """
+        try:
+            enrollment = CourseEnrollment.enroll(
+                user=user,
+                course_key=course_run_key,
+                mode=entitlement.mode,
+                check_access=True
+            )
+        except AlreadyEnrolledError:
+            enrollment = CourseEnrollment.get_enrollment(user, course_run_key)
+            if enrollment.mode == entitlement.mode:
+                CourseEntitlement.set_enrollment(entitlement, enrollment)
+            # Else the User is already enrolled in another Mode and we should
+            # not do anything else related to Entitlements.
+        except CourseEnrollmentException:
+            message = (
+                'Course Entitlement Enroll for {username} failed for course: {course_id}, '
+                'mode: {mode}, and entitlement: {entitlement}'
+            ).format(
+                username=user.username,
+                course_id=course_run_key,
+                mode=entitlement.mode,
+                entitlement=entitlement.uuid
+            )
+            return Response(
+                status=status.HTTP_400_BAD_REQUEST,
+                data={'message': message}
+            )
+
+        CourseEntitlement.set_enrollment(entitlement, enrollment)
+
+    def _unenroll_entitlement(self, entitlement, course_run_key, user):
+        """
+        Internal method to handle the details of Unenrolling a User in a Course Run.
+        """
+        CourseEnrollment.unenroll(user, course_run_key, skip_refund=True)
+        CourseEntitlement.set_enrollment(entitlement, None)
```

Comment:
Shouldn't unenroll take care of this itself?

Reply:
I chose to not mess with the CourseEnrollment code, as it is specific to Course Runs.  This separates the two. I can change this if we are concerned.